fix: spawning new player does not remove previous

NoelStephensUnity opened the pull request:
Spawning a new player object with SpawnAsPlayerObject does not set the IsPlayerObject property of the previous player object to false allowing for multiple player objects which shouldn't be the case.
[MTT-1570](https://jira.unity3d.com/browse/MTT-1570)

## Changelog
### com.unity.netcode.gameobjects
-Fixed: issue when spawning new player if an already existing player exists it does not remove IsPlayer from the previous player.

## Testing and Documentation
* Includes integration tests.

NoelStephensUnity committed:
fix

This will un-mark any NetworkObjects currently marked as a player if a new player instance is spawned for an already connected player.
Handles both client and server sides.

NoelStephensUnity committed:
test

This is the test to validate this fix.

NoelStephensUnity committed:
Update CHANGELOG.md

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -494,6 +494,12 @@ private void SpawnNetworkObjectLocallyCommon(NetworkObject networkObject, ulong
                 {
                     if (playerObject)
                     {
+                        // If there was an already existing player object for this player, then mark it as no longer
```
I think for clarity, since `(NetworkManager.ConnectedClients[ownerClientId.Value].PlayerObject` now shows up 3 times in as many lines, I'd hoist this as a local variable

mattwalsh-unity commented on com.unity.netcode.gameobjects/Tests/Runtime/PlayerObjectTests.cs:
```diff
@@ -0,0 +1,41 @@
+        [UnityTest]
+        public IEnumerator SpawnAnotherPlayerObject()
+        {
+            var originalPlayer = m_PlayerNetworkObjects[1][1];
```
I'm trying to rationalize to myself that I should know why I am looking this up by [1][1].  Worth a comment to explain.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Tests/Runtime/PlayerObjectTests.cs:
```diff
@@ -0,0 +1,41 @@
+            var newPlayerNetworkObject = newPlayer.GetComponent<NetworkObject>();
+            newPlayerNetworkObject.NetworkManagerOwner = m_ServerNetworkManager;
+            newPlayerNetworkObject.SpawnAsPlayerObject(1);
+            yield return WaitForConditionOrTimeOut(() => !originalPlayer.IsPlayerObject && newPlayerNetworkObject.IsPlayerObject);
```
Worth a comment for those who won't have this in their brain cache to say "Spawn another player object.  We should now see the original player object no longer think it's a player object",

mattwalsh-unity commented on com.unity.netcode.gameobjects/Tests/Runtime/PlayerObjectTests.cs:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        [UnityTest]
+        public IEnumerator SpawnAnotherPlayerObject()
```
I would title this "TestReplacementPlayerObjectCleanup" or similar.  "SpawnAnotherPlayerObject", if I'm coming in cold, just tells me we are testing spawning yet another PlayerObject - a title that specifically conveys that you are checking that old player objects are de-player-ified will help future generations

mattwalsh-unity reviewed (changes requested):
Would like some more comments in the test and have the test named more self-describingly

NoelStephensUnity committed:
update

updating name for better clarity
removed hard-coded indices values with relative properties
adding comments for clarity

NoelStephensUnity committed:
style

updated comments slightly.

NoelStephensUnity committed:
style

minor typo adjustment
